### PR TITLE
construct evm circuit test inputs from geth trace

### PR DESCRIPTION
**Pull request by HAOYUatHZ**
1. use geth trace to construct evm circuit test inputs (in order to achieve this, we add a `bus_mapping_tmp_convert` mod)
2. move `bus_mapping_tmp` and `bus_mapping_tmp_convert` to `witness` mod.
3. `bus_mapping::mstore` can support both `mstore` and `mstore8` now.

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/witness.rs [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+    }
+}
+
+use halo2::arithmetic::BaseExt;
+use pairing::bn256::Fr as Fp;
+use std::convert::TryInto;
+
+impl From<&bus_mapping::circuit_input_builder::ExecStep> for ExecutionState {
+    fn from(step: &bus_mapping::circuit_input_builder::ExecStep) -> Self {
+        // TODO: convert error (circuit_input_builder.rs)
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/witness.rs:322]: rephrase

File: zkevm-circuits/src/evm_circuit/witness.rs [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+        rw_counter: usize::from(step.gc),
+        program_counter: usize::from(step.pc) as u64,
+        stack_pointer: 1024 - step.stack_size,
+        gas_left: step.gas_left.0,
+        gas_cost: step.gas_cost.as_u64(),
+        opcode: Some(step.op),
+        // geth increases memory size before making trace,
+        // so the memory size in ExecStep is not what we expect to see.
+        // We have to use the memory size of previous step as correct memory
+        // size before each step
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/witness.rs:397]: Add link to geth: https://github.com/ethereum/go-ethereum/blob/master/core/vm/interpreter.go#L224-L233 in comments

File: zkevm-circuits/src/evm_circuit/execution/memory.rs [hunk 3]
```diff
@@ -201,117 +199,52 @@ impl<F: FieldExt> ExecutionGadget<F> for MemoryGadget<F> {
...
+        let bytecode = bytecode! {
+            PUSH32(value)
+            PUSH32(address)
+            #[start]
+            .write_op(opcode)
+            STOP
         };
+        //let block  =
+        // witness::build_block_from_trace_code_at_start(&
+        // bytecode);
```
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/memory.rs:228]: clean

File: bus-mapping/src/evm/opcodes/mstore.rs [hunk 4]
```diff
@@ -12,9 +12,9 @@ use core::convert::TryInto;
 /// corresponding to the [`OpcodeId::MSTORE`](crate::evm::OpcodeId::MSTORE)
 /// `OpcodeId`.
 #[derive(Debug, Copy, Clone)]
-pub(crate) struct Mstore;
+pub(crate) struct Mstore<const WHOLEWORD: bool>;
```
[the remark this answers is not in the document]
Comment (HAOYUatHZ) [reply bus-mapping/src/evm/opcodes/mstore.rs]: resolved in https://github.com/scroll-tech/zkevm-circuits/pull/30/commits/26eebd6789cdbf9f34ba744aa3a99462a65c08db

File: bus-mapping/src/evm/opcodes/stackonlyop.rs [hunk 5]
```diff
@@ -23,11 +23,12 @@ impl<const N: usize> Opcode for StackOnlyOpcode<N> {
...
         // N stack reads
         for i in 0..N {
-            state.push_op(StackOp::new(
+            let op = StackOp::new(
                 RW::READ,
                 step.stack.nth_last_filled(i),
                 step.stack.nth_last(i)?,
-            ));
+            );
+            state.push_op(op);
```
[the remark this answers is not in the document]
Comment (HAOYUatHZ) [reply bus-mapping/src/evm/opcodes/stackonlyop.rs]: reverted in c30d361f4c754f59799eaddd14610745b07cb8cc

File: zkevm-circuits/src/evm_circuit/execution/memory.rs (hunk 3 above)
Comment (HAOYUatHZ) [reply zkevm-circuits/src/evm_circuit/execution/memory.rs]: addressed in c30d361f4c754f59799eaddd14610745b07cb8cc

File: zkevm-circuits/src/evm_circuit/witness.rs (hunk 2 above)
Comment (HAOYUatHZ) [reply zkevm-circuits/src/evm_circuit/witness.rs:399]: addressed in https://github.com/scroll-tech/zkevm-circuits/pull/30/commits/3184aec2ffa9395a6b2af729c76249dd65d0a899

File: zkevm-circuits/src/evm_circuit/witness.rs (hunk 1 above)
Comment (HAOYUatHZ) [reply zkevm-circuits/src/evm_circuit/witness.rs]: addressed in https://github.com/scroll-tech/zkevm-circuits/pull/30/commits/3184aec2ffa9395a6b2af729c76249dd65d0a899

**Comment by icemelon** [comment]
great. I've sent the PR to EF: https://github.com/appliedzkp/zkevm-circuits/pull/239

**Pull request closed without merging**